module/applications/components/enchantment-application.mjs

```diff
@@ -0,0 +1,145 @@
+/**
+ * Application to handle applying enchantments to items from a chat card.
+ */
+export default class EnchantmentApplicationElement extends HTMLElement {
+
+  /* -------------------------------------------- */
+  /*  Properties                                  */
+  /* -------------------------------------------- */
+
+  /**
+   * The chat message with which this enchantment is associated.
+   * @type {ChatMessage5e}
+   */
+  chatMessage;
+
+  /* -------------------------------------------- */
+
+  /**
+   * Area where items can be dropped to enchant.
+   * @type {HTMLElement}
+   */
+  dropArea;
+
+  /* -------------------------------------------- */
+
+  /**
+   * Item providing the enchantment that will be applied.
+   * @type {Item5e}
+   */
+  get enchantmentItem() {
+    return this.chatMessage.getAssociatedItem();
+  }
+
+  /* -------------------------------------------- */
+  /*  Rendering                                   */
+  /* -------------------------------------------- */
+
+  connectedCallback() {
+    const messageId = this.closest("[data-message-id]")?.dataset.messageId;
+    this.chatMessage = game.messages.get(messageId);
+    if ( !this.chatMessage ) return;
+
+    // Build the frame HTML only once
+    if ( !this.dropArea ) {
+      const div = document.createElement("div");
+      div.classList.add("enchantment-control");
+      div.innerHTML = '<div class="drop-area"></div>';
+      this.replaceChildren(div);
+      this.dropArea = div.querySelector(".drop-area");
+      this.addEventListener("drop", this._onDrop.bind(this));
+      this.addEventListener("click", this._onRemoveEnchantment.bind(this));
+    }
+
+    this.buildItemList();
+  }
+
+  /* -------------------------------------------- */
+
+  /**
+   * Build a list of enchanted items.
+   */
+  async buildItemList() {
+    const enchantedItems = (await Promise.all(
+      (this.chatMessage.getFlag("dnd5e", "enchanted") ?? []).map(uuid => fromUuid(uuid))
```

Comment:
Is it necessary to store these on the chat message itself? Each enchantment Item knows which other Items it has enchanted, so perhaps this information can be retrieved from there? I don't think it's particularly important that each chat message remembers which Items it enchanted specifically. It should be ok when posting an enchantment to chat that you see all its enchanted items and can add/remove them still. Maybe not, but for the enchantment use cases I can think of, it seems ok. It would mean we remove the requirement for an active GM to be present to orchestrate this.